cloudflare/src/request.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{Credentials, Endpoint, CLIENT_API_V4_URI};
+use url::Url;
+
+/// A basic builder that contains credentials and a base API URL for use with constructing requests
+/// for Endpoints.
+pub struct RequestBuilder {
+    /// The credentials to use for requests
+    pub credentials: Credentials,
+    /// The API URL to make requests to
+    pub api_url: Url,
+}
+
+impl RequestBuilder {
+    /// Creates a new builder with the provided credentials and the default client API url.
+    pub fn new(credentials: Credentials) -> Self {
+        Self {
+            credentials,
+            api_url: CLIENT_API_V4_URI.parse().expect("default API URI is valid"),
+        }
+    }
+
+    /// Attempts to create a request using this builder and the specified endpoint. If the API Url
+    /// cannot be joined with the endpoint's path, this returns a URL parse error.
+    pub fn try_build<'a, E: Endpoint + 'a>(
+        &'a self,
+        endpoint: &'a E,
+    ) -> Result<Request<'a, E>, url::ParseError> {
+        Ok(Request {
+            url: self.api_url.join(&endpoint.path())?,
+            credentials: &self.credentials,
+            query: endpoint.query(),
+            body: endpoint.body(),
+        })
+    }
+
+    /// Creates a request using this builder and the specified endpoint.
+    ///
+    /// # Panics
+    ///
+    /// If the API Url cannot be joined with the endpoint's path.
+    pub fn build<'a, E: Endpoint + 'a>(&'a self, endpoint: &'a E) -> Request<'a, E> {
```

Comment:
nit: I'd prefer if the shorter name "build" was the safe one (i.e. doesn't panic), and the panicking method had a longer name than "build". Or maybe we could remove the panic, and just let users call .build().unwrap() if they really want to force the panic.

Reply:
I prefer shorter names to be obvious happy paths that most users will take. 99.99% of the time the user will have a URL that supports joining with paths (if they use the default URL this will be true) and endpoints will have paths that can be joined with a standard base URL (if they use included endpoints, this will also be true), anything else is something a user probably doesn't want to handle, but if they do, they can use the `try_` methods.